### ISSUE TITLE
CoreCLR's CNTRL_SHUTDOWN_EVENT handler prevents graceful exit of services during system shutdown

### ISSUE THREAD

**Issue by gkapellmann** [dotnet/runtime#83093]
### Description

I am not sure if this belongs here, but this is a new weird behaviour I have been fighting with lately.

I have a service, where the 'Main_Worker' is a 'IHostedService' and a 'BackgroundService'. I have overriden the 'StopAsync' function and the only thing it really does is write to the Windows Event Log a message and save a txt file with some data.

While I was working with this structure in .NET 6, it all worked preatty well. Then I upgraded to .Net 7, and it seems that now 'StopAsync' is not triggering when I turn of the PC by a normal Power -> Shutdown or if I use 'Process.Start("cmd", @"/c shutdown -s -t 0");'

When in .NET 6, both ways it worked without problems. 

### Reproduction Steps

`    public async override Task<Task> StopAsync(CancellationToken cancellationToken) {
         Log.Text("Stopping Worker_Main service...");
         logger.LogWarning("Test Service is Stopping...");
         await Log.SaveCurrentLog();
         return base.StopAsync(cancellationToken);
     }
`

That is my StopAsync function, it used to work like this with no problems. But now, even if you take away the await line, not even the logger is appearing.

### Expected behavior

At least I would expect to see the logged message in the Windows Event Logger when the PC is turned on again.

### Actual behavior

No trace of the message being logged or the file being written. The service probably closes, yet I have no idea if it is in a forced mode or if it is actually being closed in a gracefull way.

### Known Workarounds

I do not have workarounds at the moment.

### Configuration

Have tested in Windows 10 & 11
Both are x64

**Comment by gkapellmann** [comment]
As a matter of fact, just to give a bit more of context, here is a test I have been doing:

`
    lifetime.ApplicationStarted.Register(() => logger.LogInformation("ApplicationStarted Done..."));
    lifetime.ApplicationStopping.Register(() => logger.LogInformation("ApplicationStopping Done..."));
    lifetime.ApplicationStopped.Register(() => logger.LogInformation("ApplicationStopped Done..."));
`

ApplicationStarted appears in the event viewer perfectly well. But the other two are not. This when restaring or shutdown.

But if I stop gracefully the service manually, both messageds appear fine.

**Comment by ericstj** [comment]
This one is actually different than the error 1067 issue.  In that case the service process was exiting before completing ServiceBase's OnStop method which is what notified SCM that the service cleanly stopped.  In that case it would run all ApplicationStopping and ApplicationStopped.

In this case the machine is shutting down.  I was looking at a similar issue internally and found that the problem was actually due to https://github.com/dotnet/runtime/pull/41101.  During a shutdown a service will first receive a CNTRL_SHUTDOWN_EVENT.  The default is to do nothing for a service process per https://learn.microsoft.com/en-us/windows/console/handlerroutine#remarks, but that was changed for .NET processes in https://github.com/dotnet/runtime/pull/41101.

I was able to workaround this by adding  ConsoleControlHandler that cancels other handlers when seeing CNTRL_SHUTDOWN_EVENT - stopping the runtime's handler from shutting down the process and letting the service exit normally.  In .NET 6.0 you can do this with a one-liner:
```c#
    using var registration = PosixSignalRegistration.Create(PosixSignal.SIGTERM, (context) => context.Cancel = true);
```
Just make sure that registration stays alive for the entire duration of the service - I put it in the Main() routine.

We should look into adding something like this to ServiceBase -- or having the runtime try to avoid handling CNTRL_SHUTDOWN_EVENT for services.

**Comment by ericstj** [comment]
Ideally we don't do this only in ServiceBase since that wouldn't fix cases where folks were implementing services themselves without using ServiceBase.  Instead we should try to replicate whatever the OS does in the default control handler to distinguish between normal console processes and services.

**Comment by gkapellmann** [comment]
Hello @ericstj, thank you for your answers!

I have not tested your workaround in .NET 6.0, but I did test wit with a small example in .NET 7.0 and is not working for me. As a matter of fact neither the StopAsync or StartAsync functions are getting triggered.

**Comment by ericstj** [comment]
@gkapellmann if StartAsync is not triggered then you are having a problem with even starting up your service.  Are you able to share your example?

**Comment by mangod9** [comment]
@janvorli as FYI..

**Comment by gkapellmann** [comment]
@ericstj Sure, no problem, I have created another repo with my very simple service example [here](https://github.com/gkapellmann/SimpleWinService_dotnet.git).

It is based in .NET 7.0, when running you should be ableto see the events in the event manager, then shutting the PC down and then on again and I am not able to see the start async or stop async messages, but the main executeasyc events are there properly.

**Comment by ericstj** [comment]
I tried retargeting this service to `net6.0` and noticed the exact same problem, so I'm not so sure writing to the eventlog during shutdown has ever worked correctly here.

**Comment by gkapellmann** [comment]
It is interesting, I am going to look back in my repo history. 

In the end the real priority is to have SaveCurrentLog working, the event logger is just for debugging, and I guess that what you say regarding the services makes sense. I did get the saved logs before, could it be that the order of the services shutting down is semi-random?

**Comment by ericstj** [comment]
> I did get the saved logs before, could it be that the order of the services shutting down is semi-random?

It could just be timing that's machine dependent.  I'm reproducing on a clean VM without many other services so that might mean that all the other services shut down very fast.  If you happen to have a repro that shows things working consistently, I can have a look and try to explain the difference if I can repro in my environment.  Another thing you could do to see if you were getting lucky with timing is to take your working repro and add delay (for instance, in StopAsync before calling the logger) to see if it will fail.

Probably we should add documentation to the EventLog log provider that it's not safe to use during shutdown and we might want to go and add catch statements around calls to the logger that happen during shutdown.  That feels like a pitfall in that logger cc @dotnet/area-extensions-logging

**Comment by gkapellmann** [comment]
Hello @ericstj 

I have been a bit busy but managed to do some tests. I guess I was getting lucky, I did a complete cleanup of the service and re-installed. Seems to behave different with every instalation.

But then, what I have tried is to do a loop that awaits for the stoppingToken from ExecuteAsync, then try to do a EventLog, followed by a "SaveCurrentLog" action. But I dont get a log file either here. 

There are multiple issues here: 
1.- For some reason when running as a service, the log txt file is not being written, but when running as a normal app (double click), the log txt file is written.
2.- In the image below and in the txt file log you can see the second problem. Execute Async doesnt finish totally by the time that StopAsync happens. Not allowing for the last section of EsecuteAsync to execute? I am a bit confused of what is the expected workflow of these events tbh.

![image](https://user-images.githubusercontent.com/7535220/229628427-7d996ecd-2c3e-49da-a53e-1c6b4494f595.png)

In my head I see StartAsync first, followed by ExecuteAsync. When the stopping is requested, StopAsync is triggered, but that shouldnt prevent ExecuteAsync to finish completely, right?

I have done this updates in the repo, if you want to test this changes.

**Comment by ericstj** [comment]
> When the stopping is requested, StopAsync is triggered, but that shouldnt prevent ExecuteAsync to finish completely, right?

Double check the directory.  Services run with a working directory of `c:\windows\system32`.

> When the stopping is requested, StopAsync is triggered, but that shouldnt prevent ExecuteAsync to finish completely, right?

ExecuteAsync is [called by StartAsync](https://github.com/dotnet/runtime/blob/d0913fce8975849f90fd232698d16c3f95d89d6d/src/libraries/Microsoft.Extensions.Hosting.Abstractions/src/BackgroundService.cs#L46) which receives a cancellationToken that will be cancelled when the application is stopped.

https://github.com/dotnet/runtime/blob/d0913fce8975849f90fd232698d16c3f95d89d6d/src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs#L60-L71

There is no guarantee that StopAsync waits for any completion of ExecuteAsync.  It looks to me like your ExecuteAsync method is just polling the cancellation token - so no cancellation registration is guaranteed to run --  and nothing else would wait on ExecuteAsync exiting gracefully.

**Comment by gkapellmann** [comment]
Right, so I think that things have gone really clearer for me with this.

Part of the problem is that I was trying to use the backgroundservice in a way its not meant to be used.

So:
- If ExecuteAsync has a loop, it not necessarily will be fully finished when the service is ending.
- StopAsync is called at the end, but whatever happens in it might already be unloaded or unavailable.
- StopAsync shouldn't have long tasks in it.

I have modified how I do the logging, and seems to be working fine now.

Thanks a lot @ericstj for your help through this.

If we all are ok, I am good to close this issue.

**Comment by ericstj** [comment]
This issue did help us discover the shutdown problem with CoreCLR's CNTRL_SHUTDOWN_EVENT which I mentioned here: https://github.com/dotnet/runtime/issues/83093#issuecomment-1470814801.

We've seen other reports of this same problem, so I'd like to keep this issue open to track a fix for that in CoreCLR.  @janvorli was looking into a fix for that.

**Comment by mangod9** [comment]
believe there is a workaround in place to help with this issue? Will mark it for 9 to determine the feasibility of the fix.

**Comment by ericstj** [comment]
The workaround is required in the customer's application, and they would need to identify this problem (exceptions not being caught/stop not completing on shutdown) and understand what was going on to know they needed the workaround - which is a pretty big stretch given how hard it is to debug shutdown issues.  We would still like to see a fix for this.

**Comment by mangod9** [comment]
Ok, chatted with @janvorli on this to follow up if there is a good solution on the runtime side.

**Comment by jkotas** [comment]
@janvorli @ericstj What do you expect the runtime fix to be?

If I am interpreting the comments above correctly, this is what's happening in general:
1. The OS starts sending shutdown callbacks to services (e.g. `CONTROL_SHUTDOWN`).
2. The service does not cleanup completely before returning from these callbacks. There is some ongoing async/background activity.
3. This background activity can encounter unhandled exceptions if it attempts to call into other services that has been shutdown already (e.g. log events using EventLog)

To avoid unhandled exceptions in the last step, I think the service either needs to shutdown completely in the response to the shutdown callback and/or the service needs to start ignoring unhandled exceptions at some point during shutdown. https://github.com/dotnet/runtime/issues/42275 may help with the later.

**Comment by janvorli** [comment]
The primary issue is that a service should not call ExitProcess when receiving the CTRL_SHUTDOWN_EVENT, as described in https://learn.microsoft.com/en-us/windows/console/handlerroutine#remarks:
> When a console application is run as a service, it receives a modified default console control handler. This modified handler does not call [ExitProcess](https://learn.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-exitprocess) when processing the CTRL_LOGOFF_EVENT and CTRL_SHUTDOWN_EVENT signals. This allows the service to continue running after the user logs off. If the service installs its own console control handler, this handler is called before the default handler. If the installed handler calls ExitProcess when processing the CTRL_LOGOFF_EVENT signal, the service exits when the user logs off.

The problem got introduced when .NET started to register the control handler in #55333. 
To fix that, we should change the behavior of our handler when the .NET app is running as a service. In fact, I've discussed that with Windows developers in the past and they said that the MSDN doc is slightly misleading and this is true for system session, not just services.

**Comment by janvorli** [comment]
I haven't looked into the details of our implementation of the code related to the handler, maybe it would be sufficient to return FALSE from the HandlerRoutine if we detect we are running in the system session at this point: https://github.com/dotnet/runtime/blob/ac07ea6583cfa7e6896c2fe97b595f2a395ed3ea/src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/PosixSignalRegistration.Windows.cs#L83-L85

**Comment by ericstj** [comment]
@janvorli has it right

The problem is the process is getting two different shutdown signals.  Services should only be shutdown through SCM - and that's how it works in .NETFramework and earlier .NET versions.  It broke with https://github.com/dotnet/runtime/pull/55333.  The Windows default control handler identifies a service process and avoids sending it CTRL_SHUTDOWN_EVENT - the CoreCLR should do the same since it's behaving like the OS and registering a handler for all application types.  To behave most like the OS here we should try to identify if the process is a service process - we can't count on it using System.ServiceProcess - some .NET processes opt to PInvoke directly to SCM.  If possible we should replicate the OS's detection mechanism.

**Comment by jkotas** [comment]
> It broke with https://github.com/dotnet/runtime/pull/55333

#50527 and #55333 introduced a low-level signal handling API. I do not think that the introduction of these APIs alone broke anything. The break was introduced as side-effect of ConsoleLifeTime using these APIs in https://github.com/dotnet/runtime/pull/56057.

It does not make sense for this low-level signal handling API to opinionated and handle app-model specific concerns, like skipping the handler registration for Windows Services. Windows Services are not the only ones with this type of issue. For example, the Win32 GUI apps should be only shutdown through WM_QUIT message.

I think that the problem is caused by ConsoleLifeTime being used by Windows Services. Windows Services are not console apps, so it does not make sense for them to use Console app model. (And similarly, Win32 GUI apps are not console apps and so they should not use ConsoleLifeTime. Dtto for all other appmodels - WinUI, MAUI, Avalonia, ... .)

**Comment by janvorli** [comment]
It is true that the #55333 on its own without any SIGTERM registered handler doesn't cause the issue and the problem is that the `HandlePosixSignal` sets `context.Cancel=true` which was introduced by the #56057. So, it seems the fix should be rather in the `HandlePosixSignal` - to not to set context.Cancel for processes running under system session.

**Comment by janvorli** [comment]
Or I guess just not using the ConsoleLifetime. That's what you were trying to say, @jkotas, right?

**Comment by jkotas** [comment]
> Or I guess just not using the ConsoleLifetime. That's what you were trying to say, @jkotas, right?

Right. ConsoleLifetime implements console appmodel that is not compatible with Windows services.

Note that we have SystemdLifetime that handles same concern for Systemd processes ("Unix services"): https://github.com/dotnet/runtime/blob/d88c7ba88627b4b68ad523ba27cb354809eb7e67/src/libraries/Microsoft.Extensions.Hosting.Systemd/src/SystemdLifetime.netcoreapp.cs#L15-L18 . I assume that ConsoleLifetime would not work well for Systemd processes for similar reasons.

**Comment by ericstj** [comment]
Folks aren't using `ConsoleLifetime` for windows services, they are using `WindowsServiceLifetime` which doesn't use Posix registration - but instead uses SCM for these events.  I don't think there's anything wrong the user code or the hosting code is doing here.  They are expecting for the process to continue running until they can complete the SCM request to shutdown.  That's not happening because CoreCLR registered a console handler and the OS's default handler (which avoids shutting down services) is no longer used.  It's not running because of some console lifetime registration - the coreCLR registers this for **all** applications.  There is no way for an app to tell the coreCLR not do this, (other than the hack I shared which inserts another handler and avoids calling the coreCLR).

https://github.com/sbomer/runtime/blob/29bd3c4d15f0de78b845f628058ccb17d7b4eb0c/src/coreclr/src/vm/ceemain.cpp#L639
https://github.com/sbomer/runtime/blob/29bd3c4d15f0de78b845f628058ccb17d7b4eb0c/src/coreclr/src/vm/ceemain.cpp#L404-L408

**Comment by jkotas** [comment]
Makes sense. It is .NET Core -> .NET Framework regression introduced by https://github.com/dotnet/coreclr/pull/17265.

https://github.com/dotnet/coreclr/pull/17265 was attempt to fix the console lifetime issues without introducing proper signal handling APIs. It did not work well, and we ended up introducing proper signal handling APIs anyway.

Should we revert https://github.com/dotnet/coreclr/pull/17265 and go back to .NET Framework behavior so that we are not baking app-model specific concerns into runtime?

- We may need to treat it as a breaking change
- It would align the behavior between regular CoreCLR and NAOT. NAOT does not register the shutdown handler and nobody noticed so far.

**Comment by jkotas** [comment]
Fixed by https://github.com/dotnet/runtime/pull/115494
